src/parser.rs

```diff
@@ -0,0 +1,555 @@
+use std::error::Error;
+
+use crate::command::{cmd, runnable};
+use crate::errors::SyntaxError;
+use crate::pipeline::Pipeline;
+use crate::redirect::{
+    InputRedirect, OutputRedirect, OutputRedirectAppend, Redirect, RedirectType,
+};
+use crate::sequence::{AndSequence, Sequence};
+use crate::token::Token;
+use crate::traits::{Runnable, ShellCommand};
+
+// Type aliases for readability
+type ShellCommandBox = Box<dyn ShellCommand>;
+type RunnableBox = Box<dyn Runnable>;
+type ShellCommandBoxes = Vec<ShellCommandBox>;
+type RunnableBoxes = Vec<RunnableBox>;
+
+pub fn tokenize(input: String) -> Result<RunnableBox, Box<dyn Error>> {
+    let input = preprocess(input);
+    let tokens = lex(input)?;
+    debug!("Tokens: {:?}", tokens);
+    let commands = parse(tokens)?;
+    Ok(commands)
+}
+
+pub fn preprocess(input: String) -> String {
+    // Remove leading and trailing whitespace
+    let mut input = input.trim().to_string();
+
+    // Remove trailing newline
+    if input.ends_with('\n') {
+        input.pop();
+    }
+
+    // Add semicolon to trigger final command
+    if !input.ends_with(';') {
+        input.push(';');
+    }
+
+    input
+}
+
+pub fn lex(input: String) -> Result<Vec<Token>, Box<dyn Error>> {
+    let mut escaped = false;
+    let mut quote_stack = Vec::<char>::new();
+    let mut current_token = String::new();
+    let mut tokens = Vec::<Token>::new();
+
+    for c in input.chars() {
+        let meta_ignored = escaped || !quote_stack.is_empty();
+        debug!("Current char: '{}'", c);
+        match c {
+            // Meta characters
+            '&' | '<' | '>' | ';' | '|' if !meta_ignored => {
+                debug!("Token: {}", current_token);
+                if !current_token.is_empty() {
+                    tokens.push(Token::Plain(current_token.clone()));
+                    current_token.clear();
+                }
+
+                // Check for >>
+                if c == '>' && tokens.last() == Some(&Token::Meta(">".to_string())) {
+                    tokens.pop();
+                    tokens.push(Token::Meta(">>".to_string()));
+                }
+                // Push meta character as a token
+                else {
+                    tokens.push(Token::Meta(c.to_string()));
+                }
+            }
+            ' ' if !meta_ignored => {
+                debug!("Found whitespace");
+                if !current_token.is_empty() {
+                    tokens.push(Token::Plain(current_token.clone()));
+                    current_token.clear();
+                }
+            }
+            '\\' if !meta_ignored => {
+                debug!("Found backslash");
+                escaped = true;
+            }
+            '\'' | '"' if !escaped => {
```

Comment:
Rather than the stack of quotes (don't get why stack in first place, can there be nested quotes?), you can just get the entire quote token here.
Also means you won't need the escaped variable, as at it will do is delay when you exit the loop in here

Reply:
Fair point.